core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/environment/network/EnvironmentBaseNetworkConverter.java

```diff
@@ -32,34 +34,37 @@ public Network convertToLegacyNetwork(EnvironmentNetworkResponse source, String
         Map<String, Object> attributes = new HashMap<>();
-        Optional<CloudSubnet> cloudSubnet;
-        if (StringUtils.isNotBlank(source.getPreferedSubnetId())) {
-            LOGGER.debug("Choosing subnet by prefered subnet Id {}", source.getPreferedSubnetId());
-            CloudSubnet cloudSubnetById = source.getSubnetMetas().get(source.getPreferedSubnetId());
-            if (cloudSubnetById == null) {
-                cloudSubnetById = source.getSubnetMetas().values()
-                        .stream()
-                        .filter(e -> e.getId().equals(source.getPreferedSubnetId()))
-                        .findFirst().orElse(null);
-            }
-            cloudSubnet = Optional.of(cloudSubnetById);
-        } else if (StringUtils.isNotEmpty(availabilityZone)) {
-            LOGGER.debug("Choosing subnet by availability zone {}", availabilityZone);
-            cloudSubnet = source.getSubnetMetas().values().stream()
-                    .filter(s -> StringUtils.isNotEmpty(s.getAvailabilityZone()) &&
-                            s.getAvailabilityZone().equals(availabilityZone))
-                    .findFirst();
-        } else {
-            LOGGER.debug("Fallback to choose random subnet");
-            cloudSubnet = source.getSubnetMetas().values().stream().findFirst();
-        }
+        Optional<CloudSubnet> cloudSubnet = chooseSubnet(source.getPreferedSubnetId(), source.getSubnetMetas(), availabilityZone, true);
         if (cloudSubnet.isEmpty()) {
             throw new BadRequestException("No subnet for the given availability zone: " + availabilityZone);
         }
         LOGGER.debug("Chosen subnet: {}", cloudSubnet.get());
+
+        if (source.getPublicEndpointAccessGateway() == PublicEndpointAccessGateway.ENABLED) {
+            String selectedAZ = cloudSubnet.get().getAvailabilityZone();
+            Map<String, CloudSubnet> subnetsToParse;
+            if (source.getGatewayEndpointSubnetMetas() == null || source.getGatewayEndpointSubnetMetas().isEmpty()) {
+                subnetsToParse = source.getSubnetMetas();
+            } else {
+                subnetsToParse = source.getGatewayEndpointSubnetMetas();
+            }
+            Map<String, CloudSubnet> publicSubnetMetas = subnetsToParse.entrySet().stream()
+                .filter(entry -> !entry.getValue().isPrivateSubnet())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            Optional<CloudSubnet> endpointGatewayCloudSubnet = chooseSubnet(null,
+                publicSubnetMetas, selectedAZ, false);
+            if (endpointGatewayCloudSubnet.isEmpty()) {
+                throw new BadRequestException("Could not find public subnet in availability zone: " + availabilityZone);
+            }
+            LOGGER.debug("Chosen endpoint gateway subnet: {}", endpointGatewayCloudSubnet.get());
+            attributes.put("endpointGatewaySubnetId", endpointGatewayCloudSubnet.get().getId());
+        }
```

Comment:
could you move it to a new method called sth like chooseSubnetForEndpointGateway?